File: libs/pilight/hardware/mqtt.lua
--
-- Copyright (C) CurlyMo
--
-- This Source Code Form is subject to the terms of the Mozilla Public
-- License, v. 2.0. If a copy of the MPL was not distributed with this
-- file, You can obtain one at http://mozilla.org/MPL/2.0/.
--

local dump = require "dump";
local json = require "json";
local lookup = require "lookup";

local M = {}

function M.validate()
	return;
end

function M.callback(mqtt, data)
	local tmp = mqtt.getUserdata();

	if data['type'] == MQTT_CONNACK then
		mqtt.subscribe("#");
	end
	if data['type'] == MQTT_PUBLISH then
		local broadcast = pilight.table();
		broadcast['origin'] = 'receiver';
		broadcast['message'] = {};
		broadcast['message']['topic'] = data['topic'];
		broadcast['message']['payload'] = data['message'];
		broadcast['protocol'] = 'mqtt';

		local event = pilight.async.event();
		event.register(pilight.reason.RECEIVED_API);
		event.trigger(broadcast());
	end
	if data['type'] == MQTT_SUBACK then
	end
end

function M.run(a)
	local mqtt = pilight.network.mqtt();

	local config = pilight.config();
	local data = config.getData();
	local port = lookup(data, 'settings', 'mqtt-port') or 1883;

	mqtt.setCallback("callback");
	mqtt.connect("127.0.0.1", 1883);
end

function M.implements()
	return nil;
end

function M.info()
	return {
		name = "mqtt",
		version = "4.1",
		reqversion = "7.0",
		reqcommit = "94"
	}
end

return M;

File: tests/mqtt.c
/*
  Copyright (C) CurlyMo

  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#ifndef _WIN32
	#include <unistd.h>
#endif

#include "../libs/pilight/core/CuTest.h"
#include "../libs/pilight/core/pilight.h"
#include "../libs/pilight/core/mqtt.h"
#include "../libs/pilight/core/log.h"
#include "../libs/pilight/core/network.h"
#include "../libs/pilight/core/eventpool.h"
#include "../libs/libuv/uv.h"

#include "alltests.h"

void test_mqtt_decode(CuTest *tc);
void test_mqtt_encode(CuTest *tc);
void test_mqtt_server_client(CuTest *tc);
void test_mqtt_blacklist(CuTest *tc);

CuSuite *suite_mqtt(void) {
	CuSuite *suite = CuSuiteNew();

	SUITE_ADD_TEST(suite, test_mqtt_decode);
	SUITE_ADD_TEST(suite, test_mqtt_encode);
	SUITE_ADD_TEST(suite, test_mqtt_server_client);
	SUITE_ADD_TEST(suite, test_mqtt_blacklist);

	return suite;
}